content/posts/0004-writing-buildpacks.md

```diff
@@ -0,0 +1,91 @@
+![Roads-Diverged](/images/posts/0004/roads-diverged.jpg)
+
+##### When to add a new buildpack
+New functionality usually means a new buildpack is needed. An example of this can be seen with the [Rails Assets buildpack](https://github.com/paketo-buildpacks/rails-assets). A user, [@drnic](https://github.com/drnic) filed [an issue](https://github.com/paketo-buildpacks/ruby/issues/470) about performing asset compilation for Ruby apps, which was completely unsupported behaviour in the Ruby buildpacks. A Ruby maintainer, [@ryanmoran](https://github.com/ryanmoran) wrote up [an RFC](https://github.com/paketo-buildpacks/ruby/pull/475) to propose a new buildpack for the functionality, and a contributor [@genevievelesperance](https://github.com/genevieve) actually [implemented the buildpack](https://github.com/genevieve/rails-assets). (If you’re interested in this workflow, look out for a future blog post outlining our contribution process!) The idea here is that a new buildpack was needed in order to support completely new functionality.
+
+##### When to modify a buildpack
+On the other hand, there are many cases for modifying an existing buildpack. When looking for cases of “unimplemented subsets of existing functionality”, ask yourself: *If you were building this app again, is there any “optional” behaviour?*
+
+For example, imagine an app that serves content on a custom port, rather than using the default port 8080. If the existing buildpack only allowed default port 8080 to be used, it would make sense to modify the buildpack to allow for the optional behaviour of using a custom port, rather than writing a separate buildpack with almost the same logic. This is a simplified example, but there are surely other examples of optimizations within the scope of a buildpack’s functionality that could allow for some new behaviour. We’re always looking for cases where we can optimize the buildpacks, and invite you to weigh in via a Github issue if you can think of one!
+
+#### Patterns are everywhere
+
+Across the different language ecosystems, the buildpacks follow major patterns depending on the type of job that they are doing. Understanding the patterns that our buildpacks follow can make the decision whether to modify a buildpack or create a new one clearer. Especially in the case of writing a new buildpack, I recommend taking advantage of the work that’s already been done. It's far more helpful to look at what other buildpacks are doing as concrete examples than to try to architect complicated behaviours from scratch. Decide what you need your buildpack to do, and find patterns that suit your case. Here I’ve highlighted some common patterns across our buildpacks:
+
+##### Installing a distribution/tool
+
+When you need to download some tool or distribution for your app, and (usually) put it on the PATH, you may find yourself saying:
```

Comment:
```suggestion
When you need to download some tool or distribution for your app, and (usually) put it on the PATH, you may find yourself saying:

```

Reply:
This will fix the formatting so that this section looks like the other "pattern" sections.

Reply:
Struggled with formatting, thank you!!